### DISCUSSION TITLE
Invalid package dependency graph: cyclic dependency detected

### DISCUSSION THREAD

**Discussion by christianalares** [vercel/turborepo#1752] [Help] [answered]
I'm not sure if this is a bug or if I'm doing something conceptually wrong. I reckon the latter. However. I have a monorepo (turborepo) with a few apps and a few packages. Let's say two of my packages are named `pkg-a` and `pkg-b`. `pkg-a` is using code from `pkg-b` and `pkg-b` is using code from `pkg-a`. In other words their `package.json`-file depends on one another.
This gives me an error in the console saying `Invalid package dependency graph: cyclic dependency detected` when I try to run any script from the root (in my case `turbo run lint`).

I guess this is intended but I don't understand how to achieve what I want. Isn't it common that two packages uses code from one another?

Thanks in advance!

C

**Comment by mehulkar** [comment, marked as answer]
> Isn't it common that two packages uses code from one another?

I'm not sure I've seen this case before, although admittedly, I haven't worked with a LOT of monorepos. I would recommend restructuring your monorepo so you don't have a cyclical dependency! One way would be to create a `pkg-c` that has the pieces that you need for one of the packages.

**Reply by Jorundur** [reply]
Absolutely wild that this has so many downvotes when it's the **only** correct answer. A cyclic dependency is by definition something like described in the initial question where A imports from B which imports from A which imports from B which...

The only way to fix that is to _break the cycle_ and move relevant parts to C... so that you end up with something like this:

- A imports from C
- A might also import from B
- B imports from C

**Reply by christianalares** [reply]
I agree, this one is almost one year old but the solution has not changed and shouldn't have. This is correct, and the way to structure your packages
